src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs

```diff
@@ -267,23 +272,51 @@ private void EnsureServer()
             }
         }
 
-        public void Dispose()
+        public async Task StopAsync(CancellationToken cancellationToken)
         {
+            if (_stopped)
+            {
+                return;
+            }
+            _stopped = true;
+
             _logger?.Shutdown();
 
+            if (!cancellationToken.CanBeCanceled)
+            {
+                cancellationToken = new CancellationTokenSource(Options.ShutdownTimeout).Token;
+            }
+
             // Fire IApplicationLifetime.Stopping
             _applicationLifetime?.StopApplication();
 
+            try
+            {
+                await Server?.StopAsync(cancellationToken);
+            }
+            catch (Exception ex)
+            {
+                _logger?.ServerShutdownException(ex);
```

Comment:
Are we OK with not rethrowing here? The sources of Kestrel crashing bugs currently get surfaced to Program.Main by throwing from IServer.Dispose, but this will likely change to throw from StopAsync. If the _logger isn't well configured, the app developer might never see why the process crashed.